# Single-tailed test

In [1]:
import pandas as pd
from scipy.stats import t as t_student
import numpy as np
from statsmodels.stats.weightstats import DescrStatsW

## <font color='red'>Problema</font>

Nossa fábrica de vinho agora está produzindo mii garrafas de 350 ml de seu principal produto que contém, no máximo, 37 gramas de açúcar. Esta alegação nos leva a entender que a quantidade média de açúcar em uma garrafa deve ser igual ou menor que 37 g.

Um consumidor desconfiado e com conhecimentos em inferência estatística resolve testar a alegação do fabricante e seleciona, aleatóriamente, em um conjunto de estabelecimentos distintos, uma amostra de 25 garrafas. Utilizando o equipamento correto o consumidor obteve as quantidades de açúcar em todas as 25 amostras.

Assumindo que essa população se distribua aproximadamente como uma normal e considerando um nível de significância de 5%, é possível aceitar como válida a alegação do fabricante?

## Student's t-distribution

### building the table

In [2]:
t_student_table = pd.DataFrame(
                      [],
                      index=[i for i in range(1,31)],
                      columns=[i / 100 for i in range(10,0,-1)]
                      )

for index in t_student_table.index:
    for column in t_student_table.columns:
        t_student_table.loc[index, column] = \
            t_student.ppf(1 - float(column) / 2, index)

index = [('Graus de Liberdade (n - 1)', i) for i in range(1, 31)]
t_student_table.index = pd.MultiIndex.from_tuples(index)

columns = [('{0:0.3f}'.format(i / 100),
           '{0:0.3f}'.format((i / 100) / 2))
           for i in range(10, 0, -1)]
t_student_table.columns = pd.MultiIndex.from_tuples(columns)

t_student_table.rename_axis(['Two-tailed', 'Single-tailed'],
                            axis=1, inplace=True)

t_student_table

Two-tailed                       0.100    0.090    0.080    0.070    0.060  \
Single-tailed                    0.050    0.045    0.040    0.035    0.030   
Graus de Liberdade (n - 1) 1   6.31375  7.02637  7.91582  9.05789  10.5789   
                           2   2.91999  3.10398  3.31976  3.57825  3.89643   
                           3   2.35336  2.47081  2.60543   2.7626  2.95051   
                           4   2.13185   2.2261  2.33287  2.45589  2.60076   
                           5   2.01505  2.09784  2.19096  2.29739  2.42158   
                           6   1.94318   2.0192  2.10431  2.20106  2.31326   
                           7   1.89458  1.96615  2.04601  2.13645  2.24088   
                           8   1.85955  1.92799  2.00415  2.09017  2.18915   
                           9   1.83311  1.89922  1.97265  2.05539  2.15038   
                           10  1.81246  1.87677   1.9481  2.02833  2.12023   
                           11  1.79588  1.85877  1.92843  2.00666  2.09614   
                           12  1.78229  1.84402  1.91231  1.98893  2.07644   
                           13  1.77093   1.8317  1.89887  1.97416  2.06004   
                           14  1.76131  1.82127   1.8875  1.96166  2.04617   
                           15  1.75305  1.81232  1.87774  1.95094  2.03429   
                           16  1.74588  1.80455  1.86928  1.94165    2.024   
                           17  1.73961  1.79776  1.86187  1.93353    2.015   
                           18  1.73406  1.79175  1.85534  1.92636  2.00707   
                           19  1.72913  1.78642  1.84953  1.91999  2.00002   
                           20  1.72472  1.78164  1.84433  1.91429  1.99371   
                           21  1.72074  1.77734  1.83965  1.90916  1.98804   
                           22  1.71714  1.77345  1.83542  1.90452  1.98291   
                           23  1.71387  1.76991  1.83157  1.90031  1.97825   
                           24  1.71088  1.76667  1.82805  1.89646  1.97399   
                           25  1.70814  1.76371  1.82483  1.89293   1.9701   
                           26  1.70562  1.76098  1.82186  1.88968  1.96651   
                           27  1.70329  1.75847  1.81913  1.88669   1.9632   
                           28  1.70113  1.75613  1.81659  1.88391  1.96014   
                           29  1.69913  1.75397  1.81424  1.88134  1.95729   
                           30  1.69726  1.75195  1.81205  1.87894  1.95465   

Two-tailed                       0.050    0.040    0.030    0.020    0.010  
Single-tailed                    0.025    0.020    0.015    0.010    0.005  
Graus de Liberdade (n - 1) 1   12.7062  15.8945  21.2049  31.8205  63.6567  
                           2   4.30265  4.84873  5.64278  6.96456  9.92484  
                           3   3.18245  3.48191  3.89605   4.5407  5.84091  
                           4   2.77645  2.99853  3.29763  3.74695  4.60409  
                           5   2.57058  2.75651  3.00287  3.36493  4.03214  
                           6   2.44691  2.61224  2.82893  3.14267  3.70743  
                           7   2.36462  2.51675  2.71457  2.99795  3.49948  
                           8     2.306  2.44898  2.63381  2.89646  3.35539  
                           9   2.26216  2.39844   2.5738  2.82144  3.24984  
                           10  2.22814  2.35931  2.52748  2.76377  3.16927  
                           11  2.20099  2.32814  2.49066  2.71808  3.10581  
                           12  2.17881  2.30272   2.4607    2.681  3.05454  
                           13  2.16037   2.2816  2.43585  2.65031  3.01228  
                           14  2.14479  2.26378   2.4149  2.62449  2.97684  
                           15  2.13145  2.24854  2.39701  2.60248  2.94671  
                           16  2.11991  2.23536  2.38155  2.58349  2.92078  
                           17  2.10982  2.22385  2.36805  2.56693  2.89823  
                           18  2.10092  

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img004.png' width='250px'>

As células da tabela acima são valores de $t$ para uma área ou probabilidade na cauda superior da distribuição $t$.

Os **testes unicaudais** verificam as variáveis em relação a um piso ou a um teto e avaliam os valores máximos ou mínimos esperados para os parâmetros em estudo e a chance de as estatísticas amostrais serem inferiores ou superiores a dado limite.

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img008.png' width='700px'>

## Problem data

In [3]:
sample = [37.27, 36.42, 34.84, 34.60, 37.49, 
          36.53, 35.49, 36.90, 34.52, 37.30, 
          34.99, 36.55, 36.29, 36.06, 37.42, 
          34.47, 36.70, 35.86, 36.80, 36.92, 
          37.04, 36.39, 37.32, 36.64, 35.45]

sample_df = pd.DataFrame(sample, columns=['Sample'])
sample_df

Sample
0    37.27
1    36.42
2    34.84
3    34.60
4    37.49
5    36.53
6    35.49
7    36.90
8    34.52
9    37.30
10   34.99
11   36.55
12   36.29
13   36.06
14   37.42
15   34.47
16   36.70
17   35.86
18   36.80
19   36.92
20   37.04
21   36.39
22   37.32
23   36.64
24   35.45

In [4]:
sample_mean = sample_df.mean()

sample_std = sample_df.std()

In [5]:
# from problem statement
mean = 37
significance = 0.05
confidence = 1 - significance
n = 25
degrees_of_freedom = n - 1

#### <font color='red'>Lembre-se, a hipótese nula sempre contém a alegação de igualdade</font>

### $H_0: \mu \leq 37$

### $H_1: \mu > 37$

## Step 2: Choosing the appropriate sample distribution

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img003.png' width=70%>

## Step 3: Fixing the test significance ($\alpha$)

In [6]:
t_student_table[22:25]

Two-tailed                       0.100    0.090    0.080    0.070    0.060  \
Single-tailed                    0.050    0.045    0.040    0.035    0.030   
Graus de Liberdade (n - 1) 23  1.71387  1.76991  1.83157  1.90031  1.97825   
                           24  1.71088  1.76667  1.82805  1.89646  1.97399   
                           25  1.70814  1.76371  1.82483  1.89293   1.9701   

Two-tailed                       0.050    0.040    0.030    0.020    0.010  
Single-tailed                    0.025    0.020    0.015    0.010    0.005  
Graus de Liberdade (n - 1) 23  2.06866  2.17696  2.31323  2.49987  2.80734  
                           24   2.0639  2.17154  2.30691  2.49216  2.79694  
                           25  2.05954  2.16659  2.30113  2.48511  2.78744

- Pay attention in the line 24
- The value in this case is 1.71088

## Getting $t_{\alpha}$

without use the t_student_table

In [7]:
t_alpha = t_student.ppf(confidence, degrees_of_freedom)

t_alpha

1.7108820799094275

![Região de Aceitação](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img009.png)

## Step 4: calculation of the test statistic and verification of this value with the acceptance and rejection areas of the test

# $$t = \frac{\bar{x} - \mu_0}{\frac{s}{\sqrt{n}}}$$

In [8]:
t = (sample_mean - mean) / (sample_std / np.sqrt(n))
t

Sample   -3.876893
dtype: float64

![Estatística-Teste](https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img010.png)

As the $t$ is on the acceptance area, then the null hypothesis is valid.

## Step 5: Acceptance or rejection of the null hypothesis

<img src='https://caelum-online-public.s3.amazonaws.com/1229-estatistica-parte3/01/img013.png' width=90%>

### <font color='red'>Critério do valor crítico</font>

> ### Teste Unicaudal Superior
> ### Rejeitar $H_0$ se $t \geq t_{\alpha}$

In [9]:
t >= t_alpha

Sample    False
dtype: bool

As we see in the figure, t is on acceptence area. So t < t_alpha, which means accept the null hypothesis.

### <font color='green'> Conclusion: With a 95% confidence level, we cannot reject H0, that is, the manufacturer's claim is true. </font>

### <font color='red'>Critério do valor $p$</font>

> ### Teste Unicaudal Superior
> ### Rejeitar $H_0$ se o valor $p\leq\alpha$

In [10]:
p_value = t_student.sf(t,df=24)

p_value

array([0.99964062])

In [11]:
p_value <= significance

array([False])

As p_value > significance, then we accept the null hypothesis.

## A simple way to do the single-tailed test 

In [12]:
test = DescrStatsW(sample_df)

In [13]:
test.ttest_mean(value=mean, alternative='larger')

(array([-3.87689312]), array([0.99964062]), 24.0)

In [14]:
t, p_value, df = test.ttest_mean(value=mean, alternative='larger')

In [15]:
p_value

array([0.99964062])